MetaMask/metamask-extension
# Fix migration 77

## Explanation

Full explanation here https://github.com/MetaMask/metamask-extension/blob/3cf251761e65b17bf9921f2e66fbf658d4b22729/app/scripts/migrations/077-supplements/077-supplements.md

In summary:

- A bug that was present a long time ago could cause `TokenListController.tokensChainsCache[chainId].data` properties to be false, which could cause a `TypeError#1: MetaMask Migration Error #77: Cannot convert undefined or null to object`, which would result in migration 77 and higher migrations not to run.
- These migrations not running could cause various bugs, but we only discovered this problem as of v10.34.0, when the bug caused was that the app would crash on load, showing the user only a simple error screen.
- To fix this we have to fix migration 77. However, once fixed, the higher migrations would run.
- This PR fixes migration 77 to handle the undefined data, but also deletes parts of  state that could overwrite important data if present in some of the higher migrations.
- This deletion of data is done via functions defined in a new set of files and imported into the 077.js file

Click the link at top for a fuller explanation 

## Manual Testing Steps

1. Install v10.0.0 and onboard
2. Upgrade to v10.23.0
3. Add a few networks and a few tokens on each network
4. Upgrade to v10.33.0
5. The networks you added before are likely gone, Add more networks, and add some nfts
6. Upgrade to v10.34.0 and try to load the app, it will crash
7. `yarn dist` the branch of this PR, and upgrade to that build
8. You should see the tokens you added in step 3, and the networks and nfts you added in step 5. You should be on the same network you were on just before upgrading to v10.34.0. There should not be any duplicates of any tokens or nfts. You should be able to open the network selector, add more networks and otherwise use metamask normall

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
